core/dropdowndiv.js

```diff
@@ -0,0 +1,519 @@
+  // Chrome may choose to wait and apply them all at once.
+  // Since we want the translation to initial X, Y to be immediate,
+  // and the translation to final X, Y to be animated,
+  // we saw problems where both would be applied after animation was turned on,
+  // making the dropdown appear to fly in from (0, 0).
+  // Using both `left`, `top` for the initial translation and then `translate`
+  // for the animated transition to final X, Y is a workaround.
+
+  Blockly.DropDownDiv.positionInternal_(
+      metrics.initialX, metrics.initialY,
+      metrics.finalX, metrics.finalY);
+  return metrics.arrowAtTop;
+};
+
+/**
+ * Helper to position the drop-down and the arrow, maintaining bounds.
+ * See explanation of origin points in Blockly.DropDownDiv.show.
+ * @param {number} primaryX Desired origin point x, in absolute px
+ * @param {number} primaryY Desired origin point y, in absolute px
+ * @param {number} secondaryX Secondary/alternative origin point x, in absolute px
+ * @param {number} secondaryY Secondary/alternative origin point y, in absolute px
+ * @returns {Object} Various final metrics, including rendered positions for drop-down and arrow.
+ */
+Blockly.DropDownDiv.getPositionMetrics = function(primaryX, primaryY, secondaryX, secondaryY) {
+  var div = Blockly.DropDownDiv.DIV_;
+  var boundPosition = Blockly.DropDownDiv.boundsElement_.getBoundingClientRect();
+
+  var boundSize = goog.style.getSize(Blockly.DropDownDiv.boundsElement_);
+  var divSize = goog.style.getSize(div);
+
+  // First decide if we will render at primary or secondary position
+  // i.e., above or below
+  // renderX, renderY will eventually be the final rendered position of the box.
+  var renderX, renderY, renderedSecondary, renderedTertiary;
+  // Can the div fit inside the bounds if we render below the primary point?
+  if (primaryY + divSize.height > boundPosition.top + boundSize.height) {
+    // We can't fit below in terms of y. Can we fit above?
+    if (secondaryY - divSize.height < boundPosition.top) {
+      // We also can't fit above, so just render at the top of the screen.
+      renderX = primaryX;
+      renderY = 0;
+      renderedSecondary = false;
+      renderedTertiary = true;
+    } else {
+      // We can fit above, render secondary
+      renderX = secondaryX;
+      renderY = secondaryY - divSize.height - Blockly.DropDownDiv.PADDING_Y;
+      renderedSecondary = true;
+    }
+  } else {
+    // We can fit below, render primary
+    renderX = primaryX;
+    renderY = primaryY + Blockly.DropDownDiv.PADDING_Y;
+    renderedSecondary = false;
+  }
+  // First calculate the absolute arrow X
+  // This needs to be done before positioning the div, since the arrow
+  // wants to be as close to the origin point as possible.
+  var arrowX = renderX - Blockly.DropDownDiv.ARROW_SIZE / 2;
+  // Keep in overall bounds
+  arrowX = Math.max(boundPosition.left, Math.min(arrowX, boundPosition.left + boundSize.width));
```

Comment:
Should the min call be "min(arrowX, boundPosition.left + boundSize.width - (divsize+Blockly.DropDownDiv.ARROW_SIZE) / 2)"? This would match the arrow's rightmost bound to the renderX bound calculated below.

An even better approach might be to adjust the renderX first and then base the arrowX on the already fixed renderX.